### DISCUSSION TITLE
any examples for register esm?

### DISCUSSION THREAD

**Discussion by pincman** [TypeStrong/ts-node#1798] [Q&A] [answered]
I want to run `main.ts`,the code
```js
import tsNode, { createEsmHooks } from 'ts-node';
// register(process.cwd(), 'tsconfig.json');
createEsmHooks(
    tsNode.register({
        files: true,
        transpileOnly: true,
        esm: true,
    }),
);
(async () => {
    await import('./main.ts');
})();
```
but give me an error

```
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /home/code/main.ts
```

**Comment by cspotcode** [comment, marked as answer]
Node makes installing ESM hooks at runtime impossible, the only way is via the `--loader` flag.  I'm on mobile but if you search around you may find discussions where people have written their own `--loader` but use our API to implement it.  Basically, they're doing `node --loader my-fancy-loader.js` and then in that file, they're using our API to create the hooks, possibly wrapping or customizing in some way.

**Reply by rguldener** [reply]
Thanks for answer here @cspotcode, I added a PR to make this clearer in the docs: https://github.com/TypeStrong/ts-node/pull/1863

Hope this helps others.

**Comment by pincman** [comment]
This method is really troublesome. It is perfect to run the node code directly in ts-node as before, and the speed is also good after combining with swc. Some small ts projects can even run in the production environment without compiling. The experience is almost similar to the golang experience .
However, since many class libraries (such as ora, etc.) have been forced to migrate to esm (even leaving no room), a lot of labor has been added, and I do not understand this official operation of nodejs.
Anyway thank you very much for your reply

**Reply by cspotcode** [reply]
I recommend creating an issue here: https://github.com/nodejs/loaders
